pkg/types/chain_reader.go

```diff
@@ -51,21 +75,260 @@ type BoundContract struct {
 	Pending bool
 }
 
-type Event struct {
-	ChainID           string
-	EventIndexInBlock string
-	Address           string
-	TxHash            string
-	BlockHash         string
-	BlockNumber       int64
-	BlockTimestamp    time.Time
-	CreatedAt         time.Time
-	Keys              []string
-	Data              []byte
+// Head TODO this is a general chain agnostic Head, its copied from mercury Chain Reader.
+// Mercury Chain Reader should be merged with this Chain Reader, so we won't have duplicate Head definitions.
+type Head struct {
+	Number    uint64
+	Hash      []byte
+	Timestamp uint64
+}
+
+type Sequence struct {
+	// TODO SequenceCursor, this should be a unique sequence identifier that chain reader impl. understands.
+	// This way we can retrieve past/future sequences (EVM log events) very granularly, but still hide the chain detail.
+	SequenceCursor string
+	Head
+	Data any
+}
+
+type ComparisonOperator int
+
+const (
+	Eq ComparisonOperator = iota
+	Neq
+	Gt
+	Lt
+	Gte
+	Lte
+)
+
+type SortDirection int
+
+const (
+	Asc SortDirection = iota
+	Desc
+)
+
+type SortBy interface {
+	GetDirection() SortDirection
+}
+
+type LimitAndSort struct {
+	SortBy []SortBy
+	Limit  uint64
+}
+
+func NewLimitAndSort(limit uint64, sortBy ...SortBy) LimitAndSort {
+	return LimitAndSort{SortBy: sortBy, Limit: limit}
+}
+
+type SortByTimestamp struct {
+	dir SortDirection
+}
+
+func NewSortByTimestamp(sortDir SortDirection) SortByTimestamp {
+	return SortByTimestamp{dir: sortDir}
+}
+
+func (o SortByTimestamp) GetDirection() SortDirection {
+	return o.dir
+}
+
+type SortByBlock struct {
+	dir SortDirection
+}
+
+func NewSortByBlock(sortDir SortDirection) SortByBlock {
+	return SortByBlock{dir: sortDir}
+}
+
+func (o SortByBlock) GetDirection() SortDirection {
+	return o.dir
+}
+
+type SortBySequence struct {
+	dir SortDirection
+}
+
+func NewSortBySequence(sortDir SortDirection) SortBySequence {
+	return SortBySequence{dir: sortDir}
+}
+
+func (o SortBySequence) GetDirection() SortDirection {
+	return o.dir
+}
+
+type Primitive interface {
+	Accept(visitor Visitor)
+}
+
+// QueryFilter can translate to any combination of nested OR and AND boolean expressions.
+// The base Expressions slice indicates AND logical operation over expressions, which can be primitives or nested boolean expressions.
+// eg. []Expression{primitive, primitive, BooleanExpression{AND, primitive, BooleanExpression{OR, primitive, primitive}} is
+// primitive AND primitive AND (primitive AND (primitive OR primitive)).
+type QueryFilter struct {
+	Expressions []Expression
+}
+
+// Expression contains either a Primitive or a BooleanExpression.
+type Expression struct {
+	Primitive         Primitive
+	BooleanExpression BooleanExpression
+}
+
+func (expr Expression) IsPrimitive() bool {
+	return expr.Primitive != nil
+}
+
+type BooleanOperator int
```

Comment:
nit/ I think `Bool` is more go like?
```suggestion
type BoolOperator int
```